### PR TITLE
FIX[dev-server]: Fix html file matching from URL

### LINKED ISSUE #9211: Dev server serves wrong page when using query parameters without extension
# 🐛 bug report
The dev server serves the wrong page when both:
* no extension is provided
* query parameters are used

## 🎛 Configuration (.babelrc, package.json, cli command)

```js
{
  	"source": [
		"index.html",
		"publish.html",
	],
}
```

## 🤔 Expected Behavior

* `/publish` returns `publish.html`
* `/publish?foo=bar` returns `pubish.html`
* `/publish.html?foo=bar` returns `publish.html`

## 😯 Current Behavior

* `/publish` correctly returns `publish.html`
* `/publish?foo=bar` **incorrectly** returns `index.html`
* `/publish.html?foo=bar` correctly returns `publish.html`

## 💻 Code Sample

Reproducible on [this commit](https://github.com/kixelated/moq-js/tree/226eeaf0bfd04a4ef34393e7a8a7e5fca86e3e51).

### PR DESCRIPTION
**Pull request by tanmoyopenroot**
Closes #9211 

# ↪️ Pull Request
In this PR we rethink how we server the HTML bundle for a given URL in the dev server.

## 💻 Examples
We get all the `htmlBundleFilePaths` and sort them based on `/`(depth)

Example:
For URL `/other` and let's say we have sorted htmlBundleFilePaths as `['/other.html', 'other/index.html' ]`
So for `/other` URL should be served with `/other.html` file. 

And if we have sorted htmlBundleFilePaths as `['other/index.html' ]`
For `/other` URL should be served with `/other/index.html` file. 

## 🚨 Test instructions
Extend the `html` example(packages/examples/HTML).

**Create multiple directories like**

Screenshot 2023-10-29 at 12 12 16 PM

**Add source and dev command** in `package.json`

```javascript
  "scripts": {
    "dev": "parcel",
    "demo": "parcel build src/index.html"
  },
  "source": [
		"src/index.html",
		"src/other.html",
		"src/temp/index.html",
		"src/temp/other.html",
                 "src/other/index.html",
		"src/other/other.html"
	],
```
Run **yarn run dev** to use the local dev server.

**Expected Behavior**

- `/other?foo=bar` correctly returns the content of `packages/examples/html/src/other.html`
- `/other.html?foo=bar` correctly returns the content of `packages/examples/html/src/other.html`
- `/other` correctly returns the content of `packages/examples/html/src/other.html`
- `/other.html` correctly returns the content of `packages/examples/html/src/other.html`
- `/other/other?foo=bar` correctly returns `packages/examples/html/src/other/other.html`
- `/other/other` correctly returns `packages/examples/html/src/other/other.html`
- `/other/index` correctly returns `packages/examples/html/src/other/index.html`

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by tanmoyopenroot** [comment]
Hey @mischnic, Need your review of this PR.

**Comment by mischnic** [comment]
The tests are failing

**Comment by mischnic** [comment]
Please don't change any existing tests, those are there for a reason and need to pass still.
But of course, add new test cases for the linked issue you're trying to fix.

**Comment by tanmoyopenroot** [comment]
> Please don't change any existing tests, those are there for a reason and need to pass still.

Screenshot 2023-12-11 at 7 08 44 AM

For the above folder :point_up_2: and If we look at this [line](https://github.com/parcel-bundler/parcel/pull/9347/files#diff-3ec2b510cd7cce447e31a86450b1eeaf573b0b8d939df04d5d06421782b3a7e3L177), We find for `/something` and expect contains of `rootIndex` - even though `/something` does not exists, which looks like wrong maybe??? For that, I wrote a different test in [line](https://github.com/parcel-bundler/parcel/pull/9347/files#diff-3ec2b510cd7cce447e31a86450b1eeaf573b0b8d939df04d5d06421782b3a7e3R185)

The same goes for the others like [this](https://github.com/parcel-bundler/parcel/pull/9347/files#diff-3ec2b510cd7cce447e31a86450b1eeaf573b0b8d939df04d5d06421782b3a7e3L429)

Just wondering if that is the expected behaviour.

**Comment by mischnic** [comment]
No, the existing tests are correct. For a SPA-style app with client-side routing, unknown routes need to return the index page.

To fix #9211, it might be enough to just remove the `?...` part before matching it with the existing logic.

**Comment by tanmoyopenroot** [comment]
> No, the existing tests are correct. For a SPA-style app with client-side routing, unknown routes need to return the index page.

Oh okay. Then I will make the required changes

**Comment by tanmoyopenroot** [comment]
> To fix #9211, it might be enough to just remove the `?...` part before matching it with the existing logic.

Added the changes. Thanks for helping me out all the way :smiley: .

**Comment by tanmoyopenroot** [comment]
Integration test is failing in `windows` environment with both node 20 and node 18

``` Failed to resolve '@parcel/transformer-js/src/esmodule-helpers.js' from 'D:/a/parcel/parcel/packages/transformers/js/src/JSTransformer.js'```

Don't think this is related to my changes.

**Comment by mischnic** [comment]
Yeah that's fine, you can ignore that. The tests are flakey.

File: packages/reporters/dev-server/src/Server.js [hunk 1]
```diff
@@ -196,6 +196,12 @@ export default class Server {
         });
 
       let indexFilePath = null;
+      let {pathname: reqURL} = url.parse(req.originalUrl || req.url);
```
Comment (mischnic) [inline packages/reporters/dev-server/src/Server.js:199]: Looks like previously it only looked at `req.url`, is there a reason why you added `originalUrl`?

I remember that we had problems with `req.originalUrl` in the past if you use [the proxy in front of the dev server](https://parceljs.org/features/development/#api-proxy).

Comment (tanmoyopenroot) [reply packages/reporters/dev-server/src/Server.js:199]: Just wanted to make sure we are using the same pathname everywhere like [here](https://github.com/parcel-bundler/parcel/pull/9347/files#diff-d23c10b6a0a9c2979d27bca0bba2e7c0eacc9f4fdc602d2c1d4681df75a95776L140).

**Review by mischnic — APPROVED** [review]

**Pull request merged**